### PR TITLE
boards: factorize stdio CDC ACM configuration

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

There is a lot copy & paste going on with boards that use CDC ACM for stdio by default.
This moves the code to a common file for samd21, nrf52 and stm32 based boards.

### Testing procedure

Testing on one board for each family should be enough

 - [x] samd21
 - [x] nrf52
 - [x] stm32

### Issues/PRs references

depends on  #15058

### CODE REVIEW

File: boards/arduino-nano-33-ble/Makefile.dep [hunk 1]
```diff
@@ -2,16 +2,5 @@ ifneq (,$(filter saul_default,$(USEMODULE)))
...
-  # Provide stdio over USB by default
-  USEMODULE += stdio_cdc_acm
-
-  # This board requires support for Arduino bootloader.
-  FEATURES_REQUIRED += bootloader_arduino
-  USEMODULE += usb_board_reset
-endif
-
-# include common nrf52 dependencies
+include $(RIOTBOARD)/common/nrf52/bootloader_nrfutil.mk
```
Comment (fjmolinas) [inline boards/arduino-nano-33-ble/Makefile.dep:5]: This is now requiring `bootloader_nrf52` instead of `bootloader_arduino`.

**Review by fjmolinas — DISMISSED** [review]
I did the following on master and this PR.

```
for board in adafruit-clue adafruit-itsybitsy-nrf52 arduino-nano-33-ble arduino-mkr1000 sodaq-one feather-m0 nrf52840dongle serpente weact-f411ce wemos-zero; do BOARD=${board} make -C examples/hello-world info-debug-variable-BOARD info-debug-variable-USEMODULE info-debug-variable-FEATURES_REQUIRED --no-print-directory; done > file.txt
```

There is the following diff:

```
diff master.txt pr-14952.txt 
9c9
< bootloader_arduino arch_32bit arch_arm cpu_core_cortexm bootloader_arduino cortexm_mpu periph_usbdev arch_32bit arch_arm cpu_core_cortexm bootloader_arduino cortexm_mpu periph_usbdev arch_32bit arch_arm cpu_core_cortexm
---
> bootloader_nrfutil arch_32bit arch_arm cpu_core_cortexm bootloader_nrfutil cortexm_mpu periph_usbdev arch_32bit arch_arm cpu_core_cortexm bootloader_nrfutil cortexm_mpu periph_usbdev arch_32bit arch_arm cpu_core_cortexm
```

File: boards/arduino-nano-33-ble/Makefile.dep (hunk 1 above)
Comment (benpicco) [reply boards/arduino-nano-33-ble/Makefile.dep]: I made it overwritable. 
What do those `bootloader_%` features actually used for? A quick `grep` shows them only being used for blacklisting certain tests entirely unrelated to the bootloader oO

File: boards/arduino-nano-33-ble/Makefile.dep [hunk 2]
```diff
@@ -2,16 +2,6 @@ ifneq (,$(filter saul_default,$(USEMODULE)))
...
-  # Provide stdio over USB by default
-  USEMODULE += stdio_cdc_acm
-
-  # This board requires support for Arduino bootloader.
-  FEATURES_REQUIRED += bootloader_arduino
-  USEMODULE += usb_board_reset
-endif
-
-# include common nrf52 dependencies
+BOOTLOADER_TYPE = bootloader_arduino
```
Comment (aabadie) [inline boards/arduino-nano-33-ble/Makefile.dep:5]:
```suggestion
# This board requires support for Arduino bootloader.
BOOTLOADER_TYPE = bootloader_arduino
```

**Review by aabadie — COMMENTED** [review]
I would rename ` boards/common/makefiles/stdio_cdc_acm.mk` in ` boards/common/makefiles/stdio_cdc_acm.dep.mk` and `boards/common/nrf52/bootloader_nrfutil.mk` in `boards/common/nrf52/bootloader_nrfutil.dep.mk`. This way it's clear that these files are about dependency resolution.

**Comment by aabadie** [comment]
I tested this PR on arduino-nano-33-ble (bossa nrf52), adafruit-clue (nrfutil) and adafruit-mkr1000 (bossa samd21): everything is working.

**Comment by benpicco** [comment]
Those `bootloader_%` features are only used to blacklist tests that fail with the CDC ACM stdio.
They are only ever used together.

How about we just add `stdio_cdc_acm` to `FEATURES_PROVIDED` instead and filter by that?
The 'bootloader' part is just misleading.

**Comment by benpicco** [comment]
I replaced that `bootloader_%` 'feature' with the proper `stdio_cdc_acm` feature now.
I wish we could just do

```
ifneq (,$(filter periph_usbdev,$(FEATURES_PROVIDED)))
  FEATURES_PROVIDED += stdio_cdc_acm
endif
```

With Kconfig, this is of course a breeze.

**Review by aabadie — CHANGES_REQUESTED** [review]
This PR is getting out of hand and now it doesn't reflect the original title and intention. It now contains debatable changes although it was ready to be merged.

Why would we had the `stdio_cdc_acm` feature but not the `stdio_uart` or `stdio_rtt` ? This question was already answered in the PR that initially introduced the arduino bootloader feature. See #12724 and #12304 where stdio over usb was initially provided as a feature (https://github.com/RIOT-OS/RIOT/pull/12304#issuecomment-541611898) but the idea was replaced by the use of `DEFAULT_MODULE`, which turns out to have his own problems.
Providing the `bootloader_%` was the cleanest approach.

Please revert the change related to the bootloader feature and keep the PR focused.

**Comment by benpicco** [comment]
The thing is that the `bootloader_%` feature is misnamed and therefore confusing.
It's only there to blacklist tests that can't work with stdio via CDC ACM and has nothing to do with the actual bootloader on the device.
I fact boards like `stm32f429i-disco` or `yarm` that don't have a bootloader would also require that bootloader feature, as those tests (no auto-init, test intentionally crashes the board) won't work there either.

So before we end up introducing new hacks (https://github.com/RIOT-OS/RIOT/pull/14952#discussion_r486522908) in trying to model the type of bootloader as a feature (that then is never used for anything), let's clean this up properly.

What

**Comment by benpicco** [comment]
Rebased on top of #15058 - this should be a lot less controversial now :wink:

**Comment by maribu** [comment]
Code looks good to me. There are to outdated requests for changes. Maybe @aabadie and @fjmolinas  could have another look now?

File: boards/common/nrf52/bootloader_nrfutil.dep.mk [hunk 3]
```diff
@@ -0,0 +1,6 @@
+include $(RIOTBOARD)/common/makefiles/stdio_cdc_acm.dep.mk
```
Comment (aabadie) [inline boards/common/nrf52/bootloader_nrfutil.dep.mk:1]: Why not move this line inside the ifneq block below ?

File: boards/common/samd21-arduino-bootloader/Makefile.dep [hunk 4]
```diff
@@ -1,5 +1,7 @@
-# This boards requires support for Arduino bootloader.
-USEMODULE += usb_board_reset
-USEMODULE += stdio_cdc_acm
+include $(RIOTBOARD)/common/makefiles/stdio_cdc_acm.dep.mk
```
Comment (aabadie) [inline boards/common/samd21-arduino-bootloader/Makefile.dep:1]: same question here.

File: boards/common/nrf52/bootloader_nrfutil.dep.mk (hunk 3 above)
Comment (benpicco) [reply boards/common/nrf52/bootloader_nrfutil.dep.mk:1]: `stdio_cdc_acm` is added to `USEMODULE` by `stdio_cdc_acm.dep.mk` (if no other stdio has been selected by the user)

So if we move it, the ifneq would always be false :wink:

File: boards/common/samd21-arduino-bootloader/Makefile.dep (hunk 4 above)
Comment (benpicco) [reply boards/common/samd21-arduino-bootloader/Makefile.dep:1]: it's independent. You can have CDC ACM without the bootloader reset, but you can't have the bootloader reset without CDC ACM.

**Review by aabadie — APPROVED** [review]
Code changes are good. I tested on sodaq-autonomo and it works. I'm confident that this will work for other boards (nrfutil based).

ACK

@fjmolinas do you agree with the latest version of this PR ?

**Pull request merged**